### PR TITLE
feat(python): adding new ast helpers

### PR DESCRIPTION
**Pull request by Dario-DC**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

Adding new ast-based helpers. Work in progress...

### CODE REVIEW

File: python/py_helpers.py [hunk 1]
```diff
@@ -99,6 +131,14 @@ def get_variable(self, name):
...
     def has_function(self, name):
         return self.find_function(name) != Node()
+    
+    def has_decorators(self, *args):
+        if not isinstance(self.tree, ast.FunctionDef):
+            return False
+        if len(args) == len(self.tree.decorator_list) \
+        and all(node.id in args for node in self.tree.decorator_list):
+            return True
+        return False
```
Comment (ojeytonwilliams) [inline python/py_helpers.py:141]:
```suggestion
        if len(args) != len(self.tree.decorator_list):
            return False
        return all(node.id in args for node in self.tree.decorator_list)
```
This reads better, I think.

File: python/py_helpers.py [hunk 2]
```diff
@@ -167,6 +207,16 @@ def find_class(self, class_name):
...
+    
+    def inherits_from(self, *args):
+        if not isinstance(self.tree, ast.ClassDef):
+            return False
+        if not self.tree.bases:
+            return False        
+        if len(args) == len(self.tree.bases) \
+        and all(node.id in args for node in self.tree.bases):
+            return True
+        return False
```
Comment (ojeytonwilliams) [inline python/py_helpers.py:219]:
```suggestion
        if len(args) != len(self.tree.bases):
            return False
        return all(node.id in args for node in self.tree.bases)
```
Same as above.

File: python/py_helpers.test.py [hunk 3]
```diff
@@ -254,6 +310,46 @@ def test_find_body_without_body(self):
...
+  def foo():
+    pass
+    
+  def bar():
+    pass    
+"""
+        node = Node(code_str)
+
+        self.assertTrue(node.find_class("A").find_function("foo").has_decorators("property", "staticmethod"))
+        self.assertFalse(node.find_class("A").find_function("foo").has_decorators("property"))
```
Comment (ojeytonwilliams) [inline python/py_helpers.test.py:350]: Is this the desired behaviour? It seems a little counter-intuitive to me, since A.foo does have `@property`

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: python/py_helpers.test.py (hunk 3 above)
Comment (Dario-DC) [reply python/py_helpers.test.py:350]: You're right, it's a bit counter-intuitive. Would it be better to check if a specific decorator (or multiple decorators) is present and avoid to check that the number of arguments passed is the same of `len(self.tree.decorator_list)`?

Comment (ojeytonwilliams) [reply python/py_helpers.test.py:350]: That makes sense to me. So, for this case I think the following should be the tests:

```py
has_decorators("property", "staticmethod") # True
has_decorators("property", "staticmethod", "anything") # False
has_decorators("property") # True
```

**Review by ojeytonwilliams — APPROVED** [review]
For future reference, Dario and I discussed using [inspect.signature](https://docs.python.org/3/library/inspect.html#inspect.Signature), but it wasn't clear how to make the api we wanted. We can revisit this if the current approach fails for some reason.

LGTM :+1:

**Pull request merged**